### PR TITLE
Make parametric_type_to_expr recursive (fixes #285)

### LINKED ISSUE #285: Warning for `LinearAlgebra.(\)` for symmetric SparseMatrixCSC{Float64,Int64} interpreted as Hermitian
I'm debugging some functions with sparse `LinearAlgebra` calls and got the following error message:
```
using LinearAlgebra, SparseArrays, Random
Random.seed!(123456)
n = 5
A = sprand(n,n,0.5)
A = A'*A
Matrix(A)
b = rand(n)
function solveit(A,b)
    println(det(A))
    return A\b
end
```
when I step into running `solveit` as follows:
```

julia> @enter solveit(A,b)
In solveit(A, b) at REPL[210]:2
>2      println(det(A))
 3      return A\b
 4  end

About to run: (LinearAlgebra.det)(<suppressed 535 bytes of output>)
1|debug> c
0.0058051455415203315
ERROR: ccall return type Ptr should have an element type (not Ptr{<:T})
Stacktrace:
 [1] allocate_sparse(::Int64, ::Int64, ::Int64, ::Bool, ::Bool, ::Int64, ::DataType) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SuiteSparse/src/cholmod.jl:474
 [2] (::DataType)(::SparseMatrixCSC{Float64,Int64}, ::Int64) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SuiteSparse/src/cholmod.jl:876
 [3] (::UnionAll)(::SparseMatrixCSC{Float64,Int64}, ::Int64) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SuiteSparse/src/cholmod.jl:910
 [4] (::UnionAll)(::Hermitian{Float64,SparseMatrixCSC{Float64,Int64}}) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SuiteSparse/src/cholmod.jl:926
 [5] #cholesky#9(::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:check,),Tuple{Bool}}}, ::typeof(cholesky), ::Hermitian{Float64,SparseMatrixCSC{Float64,Int64}}) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SuiteSparse/src/cholmod.jl:1434
 [6] (::getfield(LinearAlgebra, Symbol("#kw##cholesky")))(::NamedTuple{(:check,),Tuple{Bool}}, ::typeof(cholesky), ::Hermitian{Float64,SparseMatrixCSC{Float64,Int64}}) at none:0
 [7] \(::Hermitian{Float64,SparseMatrixCSC{Float64,Int64}}, ::Array{Float64,1}) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SuiteSparse/src/cholmod.jl:1706
 [8] \(::SparseMatrixCSC{Float64,Int64}, ::Array{Float64,1}) at /Applications/Julia-1.1.app/Contents/Resources/julia/share/julia/stdlib/v1.1/SparseArrays/src/linalg.jl:1291
 [9] solveit(::SparseMatrixCSC{Float64,Int64}, ::Array{Float64,1}) at REPL[210]:3
```
Not sure if this is an issue in `LinearAlgebra` or in the debugger. What should be done?

EDIT:
```
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
Environment:
  JULIA_LOAD_PATH = :/Users/jakeroth/Desktop/cascades/lorenz/src:/Users/jakeroth/Desktop/planning-large-deviation/code/JuPower/src:/Users/jakeroth/Desktop/planning-large-deviation/code/juthree:/Users/jakeroth/Desktop/planning-large-deviation/code/EnergyModeling/src:/Users/jakeroth/Desktop/planning-large-deviation/code/EnergyModeling/test:/Users/jakeroth/Desktop/planning-large-deviation/code/EnergyModeling/:/Users/jakeroth/Desktop/planning-large-deviation/code/EnergyFlow/src

(1.1) pkg> status
...
[31a5f54b] Debugger v0.3.0
```

### PR DESCRIPTION
**Pull request by timholy**
This doesn't work to fix #285 but it's close. The remaining problem is `JuliaInterpreter.CompiledCalls` doesn't know about other modules. Thoughts?

Relevant portion of debugging output from running the test in #285:

```julia
RetType = Ptr{SuiteSparse.CHOLMOD.C_Sparse{Tv<:Union{Complex{Float64}, Float64}}}
t.name = SuiteSparse.CHOLMOD.C_Sparse
t.name = Ptr
parametric_type_to_expr(RetType) = :(Core.Ptr{(SuiteSparse.CHOLMOD).C_Sparse{Tv}})
t.name = SuiteSparse.CHOLMOD.C_Sparse
t.name = Ptr
wrapargs = Any[:arg1, :arg2, :arg3, :arg4, :arg5, :arg6, :arg7, :arg8, :(::(Val){Tv})]
RetType = :(Core.Ptr{(SuiteSparse.CHOLMOD).C_Sparse{Tv}})
evalmod = JuliaInterpreter.CompiledCalls
def = :(function ccall_cholmod_l_allocate_sparse_libcholmod_1d562f76_7e37_11e9_2fa8_d190d118f371(arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, ::(Val){Tv}) where Tv
      #= /home/tim/.julia/dev/JuliaInterpreter/src/optimize.jl:341 =#
      return ccall(("cholmod_l_allocate_sparse", :libcholmod), Core.Ptr{(SuiteSparse.CHOLMOD).C_Sparse{Tv}}, (UInt64, UInt64, UInt64, Int32, Int32, Int32, Int32, Ptr{Nothing}), arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8)
  end)
ERROR: LoadError: UndefVarError: SuiteSparse not defined
Stacktrace:
 [1] top-level scope
 [2] top-level scope at none:0
 [3] eval at ./boot.jl:328 [inlined]
 [4] build_compiled_call!(::Expr, ::Symbol, ::Symbol, ::Array{Any,1}, ::Core.CodeInfo, ::Int64, ::Int64, ::Array{Symbol,1}, ::Module) at /home/tim/.julia/dev/JuliaInterpreter/src/optimize.jl:345
 [5] optimize!(::Core.CodeInfo, ::Method) at /home/tim/.julia/dev/JuliaInterpreter/src/optimize.jl:196
...
```

### CODE REVIEW

**Comment by timholy** [comment]
One potential workaround would be to evaluate these in the source module rather than having our own catch-all internal module. Perhaps we could then mark `Compiled()` in the `framecode`'s `methodtables`?

**Comment by timholy** [comment]
OK, this is ready for review.

**Comment by KristofferC** [comment]
I'm not gonna be able to do a deep dive review for a while but if it passes test with the additional test you put then LGTM.

File: test/interpret.jl [hunk 1]
```diff
@@ -593,3 +593,18 @@ end
...
+    function solveit(A,b)
+        return A\b .+ det(A)
+    end
+
+    Random.seed!(123456)
+    n = 5
+    A = sprand(n,n,0.5)
+    A = A'*A
+    b = rand(n)
+    @test @interpret(solveit(A, b)) == solveit(A, b)
```
Comment (KristofferC) [inline test/interpret.jl]: Does this end up getting compiled or not?

Comment (timholy) [reply test/interpret.jl]: For SuiteSparse parametric types we call the eval-fallback.

Comment (KristofferC) [reply test/interpret.jl]: Okay, they are typically expensive calls anyway so not a big deal.

**Comment by timholy** [comment]
OK, I'll merge, but obviously feel free to rework anything that bothers you at any point.

**Pull request merged**